### PR TITLE
Added IO throttling for the merger pipeline.

### PR DESCRIPTION
**Pull request by fulmicoton**
Added generic IO controls.
    
    The `IoControls` object makes it possible to
    - limit the throughput of Write and AsyncWrite object,
    - abort a write if the killswitch is activated,
    - record progress when performing IO
    - record the number of writes write in a Prometheus counter.
    
    In this PR, the IoControls object is plugged into the
    the indexer, the merger and the delete pipeline.
    
    The merge and the delete pipeline can optionally get a
    separate write throughput limit, as defined in the index
    configuration.
    
    In the merge and the delete pipeline, split downloads and
    actual index writing share the same throughput limit, and
    attempt to consume it in a concurrent manner.
    
    We rely on async-speed-limit for the throttling.
    We could not use async_speed_limit::Resource directly as
    it is not compatible with tokio.
    
    This PR also does some minor refactoring on the merge pipeline
    parameters.

throttle

### CODE REVIEW

File: quickwit/quickwit-indexing/src/actors/index_serializer.rs [hunk 1]
```diff
@@ -73,14 +73,22 @@ impl Handler<IndexedSplitBatchBuilder> for IndexSerializer {
...
-                .collect::<Result<_, _>>()?
-        };
+        let mut splits: Vec<IndexedSplit> = Vec::with_capacity(batch_builder.splits.len());
+        for split_builder in batch_builder.splits {
+            // TODO Consider & test removing this protect guard.
+            //
+            // In theory the controlled directory should be sufficient.
+            let _protect_guard = ctx.protect_zone();
+            if let Some(controlled_directory) = &split_builder.controlled_directory_opt {
+                let io_controls = controlled_directory
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/actors/index_serializer.rs:83]: You need to add the metric labels: `index_id={index_id}` and `component=index_serializer`

File: quickwit/quickwit-indexing/src/actors/index_serializer.rs [hunk 2]
```diff
@@ -73,14 +74,25 @@ impl Handler<IndexedSplitBatchBuilder> for IndexSerializer {
...
+            //
+            // In theory the controlled directory should be sufficient.
+            let _protect_guard = ctx.protect_zone();
+            if let Some(controlled_directory) = &split_builder.controlled_directory_opt {
+                let io_controls = IoControls::default()
+                    .set_progress(ctx.progress().clone())
+                    .set_kill_switch(ctx.kill_switch().clone())
+                    .set_index_and_component(
+                        &split_builder.split_attrs.pipeline_id.index_id,
+                        "index_serial",
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/actors/index_serializer.rs:89]: Nitpick: I would put `index_serializer`

File: quickwit/quickwit-indexing/src/actors/index_serializer.rs (hunk 1 above)
Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/actors/index_serializer.rs]: done

File: quickwit/quickwit-indexing/src/actors/merge_pipeline.rs [hunk 3]
```diff
@@ -238,10 +251,28 @@ impl MergePipeline {
...
+
+        let split_downloader_io_controls = IoControls::default()
+            .set_throughput_limit(max_merge_write_throughput)
+            .set_index_and_component(self.params.pipeline_id.index_id.as_str(), "splitdl_merge");
+
+        // The merge and split download share the same throughput limiter.
+        // This is how cloning the `IoControls` works.
+        let merge_executor_io_controls = split_downloader_io_controls
+            .clone()
+            .set_index_and_component(self.params.pipeline_id.index_id.as_str(), "merge");
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/actors/merge_pipeline.rs]: made it merger.

File: quickwit/quickwit-indexing/src/actors/merge_executor.rs [hunk 4]
```diff
@@ -219,78 +221,6 @@ pub fn combine_partition_ids(splits: &[SplitMetadata]) -> u64 {
     combine_partition_ids_aux(splits.iter().map(|split| split.partition_id))
 }
 
-async fn merge_split_directories(
```
Comment (fulmicoton) [inline quickwit/quickwit-indexing/src/actors/merge_executor.rs:222]: this just became a method. Diff is not very good at that, but the changed is trivial.

File: quickwit/quickwit-indexing/src/actors/merge_pipeline.rs [hunk 5]
```diff
@@ -88,15 +93,23 @@ impl Actor for MergePipeline {
...
     pub fn new(params: MergePipelineParams) -> Self {
+        let (merge_planner_mailbox, merge_planner_inbox) =
+            create_mailbox::<MergePlanner>("MergePlanner".to_string(), QueueCapacity::Unbounded);
         Self {
             params,
             previous_generations_statistics: Default::default(),
             handles: None,
             kill_switch: KillSwitch::default(),
             statistics: MergeStatistics::default(),
+            merge_planner_inbox,
```
Comment (fulmicoton) [inline quickwit/quickwit-indexing/src/actors/merge_pipeline.rs:104]: this used to be in params and I just pulled it out.

It greatly simplifies the construction.

File: quickwit/quickwit-indexing/src/actors/packager.rs [hunk 6]
```diff
@@ -134,14 +134,6 @@ impl Handler<IndexedSplitBatch> for Packager {
             split_ids=?split_ids,
             "start-packaging-splits"
         );
-        for split in &batch.splits {
-            if let Some(controlled_directory) = &split.controlled_directory_opt {
```
Comment (fulmicoton) [inline quickwit/quickwit-indexing/src/actors/packager.rs:138]: index serialization used to happen here, which is why we took over the controlled directory in the packager.

Today the serialization logic was moved int he IndexSerializer so I moved these lines to the IndexSerializer.

File: quickwit/quickwit-indexing/src/controlled_directory.rs [hunk 7]
```diff
@@ -43,36 +45,24 @@ const BUFFER_NUM_BYTES: usize = 8_192;
 /// - in the future, record a writing speed, possibly introduce some throttling, etc.
 #[derive(Clone)]
 pub struct ControlledDirectory {
-    inner: Inner,
```
Comment (fulmicoton) [inline quickwit/quickwit-indexing/src/controlled_directory.rs:46]: This inner was actually plain nonsense. It is not even in an Arc.

File: quickwit/quickwit-indexing/src/controlled_directory.rs [hunk 8]
```diff
@@ -82,77 +72,11 @@ impl fmt::Debug for ControlledDirectory {
     }
 }
 
-#[derive(Clone)]
-struct Controls {
```
Comment (fulmicoton) [inline quickwit/quickwit-indexing/src/controlled_directory.rs:86]: That logic is replaced by the ControlledWrite in quickwit_common/src/io.rs

File: quickwit/quickwit-common/src/io.rs [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+// Max 1MB at a time.
+const MAX_NUM_BYTES_WRITTEN_AT_ONCE: usize = 1 << 20;
+
+fn truncate_bytes(bytes: &[u8]) -> &[u8] {
+    let num_bytes = bytes.len().min(MAX_NUM_BYTES_WRITTEN_AT_ONCE);
+    &bytes[..num_bytes]
+}
+
+struct IoMetrics {
+    write_num_bytes_total: IntCounterVec,
```
Comment (guilload) [inline quickwit/quickwit-common/src/io.rs:56]: `num_bytes_written_total`? (`num_bytes_written` is 5x more popular on GitHub codebases)

File: quickwit/quickwit-common/src/io.rs [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+        );
+        Self {
+            write_num_bytes_total,
+        }
+    }
+}
+
+static IO_METRICS: Lazy<IoMetrics> = Lazy::new(IoMetrics::default);
+
+const REFILL_DURATION: Duration = if cfg!(test) {
```
Comment (guilload) [inline quickwit/quickwit-common/src/io.rs:75]: This variable deserves a comment.

File: quickwit/quickwit-common/src/io.rs [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+            throughput_limiter: Limiter::new(f64::INFINITY),
+            progress: Progress::default(),
+            kill_switch: KillSwitch::default(),
+            bytes_counter: default_bytes_counter,
+        }
+    }
+}
+
+impl IoControls {
+    pub fn progress(&self) -> &Progress {
```
Comment (guilload) [inline quickwit/quickwit-common/src/io.rs:103]: Yes to getters without `get_` :D

**Review by guilload — APPROVED** [review]

File: quickwit/quickwit-common/src/io.rs (hunk 9 above)
Comment (fulmicoton) [reply quickwit/quickwit-common/src/io.rs:56]: it was to stick to prometheus convention, but actually we have a unit here.

File: quickwit/quickwit-common/src/io.rs [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+        }
+        res
+    }
+}
+
+fn truncate_slices<'a, 'b>(bufs: &'b [IoSlice<'a>], max_len: usize) -> &'b [IoSlice<'a>] {
+    if bufs.is_empty() {
+        return bufs;
+    }
+    let mut cumulated_len = bufs[0].len();
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-common/src/io.rs:232]: oh yes!

[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-common/src/io.rs:232]: I fixed, added unit tests and a comment

File: quickwit/quickwit-config/src/index_config.rs [hunk 13]
```diff
@@ -72,6 +72,18 @@ pub struct IndexingResources {
     pub __num_threads_deprecated: IgnoredAny, // DEPRECATED
     #[serde(default = "IndexingResources::default_heap_size")]
     pub heap_size: Byte,
+    /// Sets the maximum write IO throughput for the merge pipeline,
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-config/src/index_config.rs:75]: Right now it is defined per index but I think this is an error. I will open an issue to address that before the release.

[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-config/src/index_config.rs:75]: tracked in #2228

File: quickwit/quickwit-common/src/io.rs (hunk 9 above)
Comment (fulmicoton) [reply quickwit/quickwit-common/src/io.rs]: rename `write_bytes`. https://prometheus.io/docs/practices/naming/

**Pull request merged**